src/backend/tests/unit/graph/graph/test_cycles.py

```diff
@@ -109,3 +114,99 @@ def test_cycle_in_graph_max_iterations():
     with pytest.raises(ValueError, match="Max iterations reached"):
         for result in graph.start(max_iterations=2, config={"output": {"cache": False}}):
             results.append(result)
+
+
+def test_that_outputs_cache_is_set_to_false_in_cycle():
+    chat_input = ChatInput(_id="chat_input")
+    router = ConditionalRouterComponent(_id="router")
+    chat_input.set(input_value=router.false_response)
+    concat_component = Concatenate(_id="concatenate")
+    concat_component.set(text=chat_input.message_response)
+    router.set(
+        input_text=chat_input.message_response,
+        match_text="testtesttesttest",
+        operator="equals",
+        message=concat_component.concatenate,
+    )
+    text_output = TextOutputComponent(_id="text_output")
+    text_output.set(input_value=router.true_response)
+    chat_output = ChatOutput(_id="chat_output")
+    chat_output.set(input_value=text_output.text_response)
+
+    graph = Graph(chat_input, chat_output)
```

Comment:
Not specific to this PR but how come you need to pass an end component to Graph? Wouldn't it eventually get added through the adjacency lists?

Reply:
Yeah. This is kinda weird still. We went for a syntax similar to Keras' Functional API. 
The end shouldn't be necessary but it has its uses still, I just didn't mind going over it yet.